### PR TITLE
P3-288 migrate revisions

### PR DESCRIPTION
**Pull request by enricobattocchi**
## Context

* When a post has been overwritten with the content of a Rewrite & Republish copy, we also want to merge the two revisions history so that the revisions of the copy (which will eventually be deleted) is not lost.

## Summary

This PR can be summarized in the following changelog entry:

* Merges the revision history of a Rewrite & Republish copy into the one of its original.

## Relevant technical choices:

* As it has happened in previous PRs, we also moved some calls to `register_hooks()` method outside the constructors for easier testing.
* Most of the main method from the new `Revisions_Migrator` class come from the `wp_save_post_revision()` function from `wp-includes/revision.php`
* The creation of a new revision on overwriting has been disabled by removing a default WP hooked function.

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Test with unlimited revisions 
* make sure you don't have `define( 'WP_POST_REVISIONS', <something> );` in your `wp-config.php`
* pick a published post and make sure it has a number of revisions.
  * On the Docker environment posts don't have revisions initially: edit a post, make a change and "Update": do it repeatedly to add revisions. 
  * I suggest to write increasing numbers in the Title and in the Content, so to "number" the revisions and see they are differing
  * take notice of the number of revisions (you can see it from the link in the sidebar: "X Revisions")
* create a Rewrite & Republish copy from that post, and do the same (small changes, like continuing the numbering in the Title and in the Content, then "Save Draft") repeatedly. Take note of the number of revisions, let's make it Y.
* Click on "Republish", then confirm if you have the pre-publish checks enabled.
* You will be redirected to the overwritten original
* Check the number of revisions: it should be X from the original, plus Y from the copy.
* You can also try to repeat the process but make some changes on the copy and *not* save them before republish: a new revision will be created before republishing thus adding +1 to the expected revisions count (and it's OK).

#### Test with limited revisions 
* make sure you have `define( 'WP_POST_REVISIONS', 5 );` in your `wp-config.php`
* pick a published post and make sure it has a number of revisions.
  * On the Docker environment posts don't have revisions initially: edit a post, make a change and "Update": do it repeatedly to add revisions. 
  * I suggest to write increasing numbers in the Title and in the Content, so to "number" the revisions and see they are differing
  * take notice of the number of revisions (you can see it from the link in the sidebar: "X Revisions")
  * try to create more than 5 revisions, and see the revisions count stays at 5.
* create a Rewrite & Republish copy from that post, and do the same (small changes, like continuing the numbering in the Title and in the Content, then "Save Draft") repeatedly. Take note of the number of revisions, let's make it Y. Better if they are less than 5.
* Click on "Republish", then confirm if you have the pre-publish checks enabled.
* You will be redirected to the overwritten original
* Check the number of revisions: it should be a total of 5.
  * inspect the DB (table `wp_post`, select the posts having `post_parent` equal to the original ID) and see that the "provenience" of the revisions is what we are expecting: the latest ones from the copy (you'll see that in the `guid` field), the earliest ones from the original, up to a total of 5.

#### Test with no revisions 
* make sure you have `define( 'WP_POST_REVISIONS', 0 );` in your `wp-config.php`
* create and publish a new post, update multiple times and see there is no revision added
* create a Rewrite & Republish copy from that post, and do the same (small changes, like continuing the numbering in the Title and in the Content, then "Save Draft") repeatedly. See there are no revisions.
* Click on "Republish", then confirm if you have the pre-publish checks enabled.
* You will be redirected to the overwritten original
* See that you still don't have any revisions
  * inspect the DB (table `wp_post`, select the posts having `post_parent` equal to the original ID) and see you get no results.
  
**remember to comment out the `define( 'WP_POST_REVISIONS', 0 );` when you have finished** so revisions are working again.

### Test instructions for QA when the code is in the RC

* [ ] QA should use the same steps as above.

QA can test this PR by following these steps:

* **Not applicable** QA will have complete test instructions for the new feature

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [X] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes [P3-288]

### CODE REVIEW

File: src/class-revisions-migrator.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		$copy          = \get_post( $copy_id );
+		$original_post = \get_post( $original_id );
+
+		if ( \is_null( $copy ) || \is_null( $original_post ) ) {
+			return;
+		}
+
+		if ( ! \wp_revisions_enabled( $original_post ) ) {
+			return;
+		}
```
Comment (JesserH) [inline src/class-revisions-migrator.php:46]: Can be done in 1 if statement

**Review by JesserH — COMMENTED** [review]
CR ✅ , if the if above comment is resolved(see comment)

File: src/class-revisions-migrator.php (hunk 1 above)
Comment (enricobattocchi) [reply src/class-revisions-migrator.php:46]: Will fix it!

**Comment by JesserH** [comment]
CR ✅

**Comment by karlijnbok** [comment]
Acceptance: 👍

**Pull request merged**